runtime/src/gpu/rocm/gpu-rocm.c

```diff
@@ -27,67 +27,497 @@
+                                        int blk_dim_x,
+                                        int blk_dim_y,
+                                        int blk_dim_z,
+                                        int nargs,
+                                        va_list args) {
+  CHPL_GPU_START_TIMER(context_time);
+
+  chpl_gpu_ensure_context();
+
+  CHPL_GPU_STOP_TIMER(context_time);
+  CHPL_GPU_START_TIMER(load_time);
+
+  hipDeviceptr_t cuda_module = chpl_gpu_cuda_modules[chpl_task_getRequestedSubloc()];
+  void* function = chpl_gpu_load_function(cuda_module, name);
+
+  CHPL_GPU_STOP_TIMER(load_time);
+  CHPL_GPU_START_TIMER(prep_time);
+
+  // TODO: this should use chpl_mem_alloc
+  void*** kernel_params = chpl_malloc(nargs*sizeof(void**));
+
+  assert(function);
+  assert(kernel_params);
+
+  CHPL_GPU_DEBUG("Creating kernel parameters\n");
+
+  // Keep track of kernel parameters we dynamically allocate memory for so
+  // later on we know what we need to free.
+  bool* was_memory_dynamically_allocated_for_kernel_param =
+    chpl_malloc(nargs*sizeof(bool));
+
+  for (int i=0 ; i<nargs ; i++) {
+    void* cur_arg = va_arg(args, void*);
+    size_t cur_arg_size = va_arg(args, size_t);
+
+    if (cur_arg_size > 0) {
+      was_memory_dynamically_allocated_for_kernel_param[i] = true;
+
+      // TODO this allocation needs to use `chpl_mem_alloc` with a proper desc
+      kernel_params[i] = chpl_malloc(1*sizeof(hipDeviceptr_t));
+
+      *kernel_params[i] = chpl_gpu_mem_alloc(cur_arg_size,
+                                             CHPL_RT_MD_GPU_KERNEL_ARG,
+                                             ln, fn);
+
+      chpl_gpu_copy_host_to_device(*kernel_params[i], cur_arg, cur_arg_size);
+
+      CHPL_GPU_DEBUG("\tKernel parameter %d: %p (device ptr)\n",
+                   i, *kernel_params[i]);
+    }
+    else {
+      was_memory_dynamically_allocated_for_kernel_param[i] = false;
+      kernel_params[i] = cur_arg;
+      CHPL_GPU_DEBUG("\tKernel parameter %d: %p\n",
+                   i, kernel_params[i]);
+    }
+  }
+
+  CHPL_GPU_STOP_TIMER(prep_time);
+  CHPL_GPU_START_TIMER(kernel_time);
+
+  CUDA_CALL(hipModuleLaunchKernel((hipFunction_t)function,
+                           grd_dim_x, grd_dim_y, grd_dim_z,
+                           blk_dim_x, blk_dim_y, blk_dim_z,
+                           0,  // shared memory in bytes
+                           0,  // stream ID
+                           (void**)kernel_params,
+                           NULL));  // extra options
+
+  CHPL_GPU_DEBUG("cuLaunchKernel returned %s\n", name);
+
+  CUDA_CALL(hipDeviceSynchronize());
+
+  CHPL_GPU_DEBUG("Synchronization complete %s\n", name);
+  CHPL_GPU_STOP_TIMER(kernel_time);
+  CHPL_GPU_START_TIMER(teardown_time);
+
+  // free GPU memory allocated for kernel parameters
+  for (int i=0 ; i<nargs ; i++) {
+    if (was_memory_dynamically_allocated_for_kernel_param[i]) {
+      chpl_gpu_mem_free(*kernel_params[i], ln, fn);
+    }
+  }
```

Comment:
need free for `was_memory_dynamically_allocated_for_kernel_param`?

Reply:
I think you're right. Looks like this is an issue in the CUDA version as well: [was_memory_dynamically_allocated_for_kernel_param](https://github.com/chapel-lang/chapel/blob/main/runtime/src/gpu/cuda/gpu-cuda.c#L216)

I created a Github issue: https://github.com/chapel-lang/chapel/issues/new

Arguably I should just "fix it" in this PR but I'd rather separate changes impacting the CUDA runtime into a separate PR.